File: frontend/src/api/planner.ts
// planner.ts
//
// Purpose
// -------
// The *only* place in this frontend that issues an HTTP request to the
// backend's Planning API (`backend/api/routes/planner.py`) -- mirrors
// `language.ts`'s "centralized API client" convention exactly. Every
// UI component calls `planTask()`/`evaluatePlanners()` from here
// instead of constructing its own `fetch()` call.
//
// Same relative-URL, no-credential rule as `language.ts`: this file
// calls `/api/v1/planner/...`, never an absolute URL, and never reads
// or forwards an LLM API key -- the backend is the only component
// that ever holds one (`backend/language/config.py`'s Security note
// applies here too, since `react` planning ultimately calls an LLM
// server-side).

import type { SingleTask } from "./types";
import type { EvaluateResponse, PlannerType, PlanningResult } from "./plannerTypes";

const DEFAULT_ERROR_MESSAGE = "The planning service could not be reached. Please try again.";

/** Thrown by this module's functions for any non-2xx response. */
export class PlannerApiError extends Error {
  readonly category: string;
  readonly status: number;

  constructor(status: number, category: string, message: string) {
    super(message);
    this.name = "PlannerApiError";
    this.status = status;
    this.category = category;
  }
}

async function postJson<T>(path: string, body: unknown): Promise<T> {
  const response = await fetch(path, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    let category = "internal_error";
    let message = DEFAULT_ERROR_MESSAGE;
    try {
      const parsed = (await response.json()) as Partial<{ category: string; message: string }>;
      if (typeof parsed.message === "string") message = parsed.message;
      if (typeof parsed.category === "string") category = parsed.category;
    } catch {
      // Response body was not JSON -- fall through to the generic message.
    }
    throw new PlannerApiError(response.status, category, message);
  }

  return (await response.json()) as T;
}

/**
 * Submits one task to `POST /api/v1/planner/plan` and returns the
 * `PlanningResult` -- always a 200 with `success`/`validation` fields
 * to check, except for a request-shape problem (unrecognized
 * `planner_type`), which throws `PlannerApiError`.
 */
export async function planTask(
  task: SingleTask,
  plannerType: PlannerType = "rule_based",
): Promise<PlanningResult> {
  return postJson<PlanningResult>("/api/v1/planner/plan", {
    task,
    planner_type: plannerType,
  });
}

/**
 * Submits one task to `POST /api/v1/planner/evaluate` and returns
 * measured metrics for every requested (default: all) planning
 * strategy -- used by the planner-comparison dashboard.
 */
export async function evaluatePlanners(
  task: SingleTask,
  plannerTypes?: PlannerType[],
): Promise<EvaluateResponse> {
  return postJson<EvaluateResponse>("/api/v1/planner/evaluate", {
    task,
    planner_types: plannerTypes,
  });
}

File: backend/api/routes/planner.py
"""
planner.py (backend/api/routes)

Purpose
-------
HTTP interface around the Phase 4 Task Planning subsystem
(`backend/planner/`). Implements three endpoints:

- `POST /api/v1/planner/plan`     -- generate a Plan for a task with a
                                      chosen strategy.
- `POST /api/v1/planner/validate` -- validate a caller-supplied Plan
                                      against a task, independent of
                                      how it was generated.
- `POST /api/v1/planner/evaluate` -- compare every (or a chosen subset
                                      of) planning strategy on the same
                                      task, returning real, freshly
                                      measured metrics (never
                                      fabricated numbers -- see
                                      `planner/evaluation.py`'s
                                      docstring).

Contains no planning logic, no validation logic, and no state
simulation of its own -- every one of those already has exactly one
owner inside `backend/planner/` (`factory.create_planner`,
`validator.PlanValidator`, `evaluation.PlannerEvaluator`). This module
only translates HTTP requests into calls against that package and
translates its results (or a `ValueError` from an unknown planner
type) into HTTP responses, mirroring `api/routes/language.py`'s own
scope boundary.

Why this router never raises for an ordinary "plan is invalid" result
-------------------------------------------------------------------------
`Planner.plan()` and `PlanValidator.validate()` both return a result
object even when the plan they describe is invalid or unachievable
-- see `planner/exceptions.py`'s docstring on why that is normal
output, not an error. This router returns those results as an
ordinary 200 response; a client checks `success`/`valid` on the body,
exactly as it already checks `ParseResponse.status` for the Language
API's "clarification_required" case. The only case this router maps to
a non-2xx status is a request-shape problem this API layer itself can
detect before ever calling into `backend/planner/`: an unrecognized
`planner_type` string (422).
"""

from __future__ import annotations

import logging

from fastapi import APIRouter, HTTPException

from api.models.planner import (
    EvaluateRequest,
    EvaluateResponse,
    PlannerMetricsResponse,
    PlanRequest,
    ValidateRequest,
)
from planner.evaluation import PlannerEvaluator
from planner.factory import available_planner_types, create_planner
from planner.models import PlanningResult, ValidationResult
from planner.state import WorldState
from planner.validator import PlanValidator

logger = logging.getLogger(__name__)

router = APIRouter()

_validator = PlanValidator()
_evaluator = PlannerEvaluator()


@router.post(
    "/plan",
    response_model=PlanningResult,
    summary="Generate a validated plan for a task",
    description=(
        "Expands a SingleTask into an ordered sequence of abstract "
        "PlanSteps using the requested planning strategy, then "
        "validates the result. Never calls AI2-THOR or produces a "
        "simulator-specific action -- see backend/planner/__init__.py."
    ),
    responses={422: {"description": "Unrecognized planner_type."}},
)
def plan_task(body: PlanRequest) -> PlanningResult:
    try:
        planner = create_planner(body.planner_type)
    except ValueError as exc:
        logger.warning(
            "planner_api.plan.invalid_planner_type",
            extra={"planner_type": body.planner_type},
        )
        raise HTTPException(
            status_code=422,
            detail={"category": "invalid_planner_type", "message": str(exc)},
        ) from None

    return planner.plan(body.task, WorldState.initial())


@router.post(
    "/validate",
    response_model=ValidationResult,
    summary="Validate a plan against a task",
    description=(
        "Replays a caller-supplied Plan against a fresh symbolic "
        "WorldState and reports structured validation issues -- "
        "independent of which strategy produced the plan."
    ),
)
def validate_plan(body: ValidateRequest) -> ValidationResult:
    return _validator.validate(body.plan, body.task, WorldState.initial())


@router.post(
    "/evaluate",
    response_model=EvaluateResponse,
    summary="Compare planning strategies on one task",
    description=(
        "Runs every requested planning strategy (default: all "
        "registered strategies) against the same task and returns "
        "measured success/validity/latency/step-count metrics for "
        "each. Metrics are always freshly measured, never fabricated."
    ),
    responses={422: {"description": "An unrecognized planner_type was requested."}},
)
def evaluate_planners(body: EvaluateRequest) -> EvaluateResponse:
    requested = body.planner_types or available_planner_types()
    try:
        planners = {name: create_planner(name) for name in requested}
    except ValueError as exc:
        raise HTTPException(
            status_code=422,
            detail={"category": "invalid_planner_type", "message": str(exc)},
        ) from None

    metrics = _evaluator.compare(planners, body.task, WorldState.initial())
    return EvaluateResponse(
        task_id=body.task.task_id or "",
        results=[
            PlannerMetricsResponse(
                planner_type=m.planner_type,
                success=m.success,
                valid=m.valid,
                goal_satisfied=m.goal_satisfied,
                latency_ms=m.latency_ms,
                step_count=m.step_count,
                invalid_action_count=m.invalid_action_count,
                redundant_step_count=m.redundant_step_count,
                error=m.error,
            )
            for m in metrics
        ],
    )
